<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Help - GLID Project</title>

    <!-- Bootstrap styles -->
    <!-- <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"> -->
    <!-- <link rel="stylesheet" href="custom.css"> -->
    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
    <!-- Optional theme -->
    <!-- <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap-theme.min.css"> -->

<style>
  section {
    border-style: solid;
    border-top-color: cornflowerblue;
    border-bottom-color: cornflowerblue;
    border-left: none;
    border-right: none;
    text-align: center;
}
</style>
  </head>
  <body>

    <nav class="navbar navbar-inverse">
      <div class="container">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.php">GLID project</a>
        </div>
        <div id="navbar" class="collapse navbar-collapse">
          <ul class="nav navbar-nav">
            <li><a href="index.php">Home</a></li>

            <li class="dropdown">
              <a href="help.html" class="dropdown-toggle" data-toggle="dropdown">Help<span class="caret"></span></a>
              <ul class="dropdown-menu">
                  <li><a href="#description">GLID description</a></li>
                  <li><a href="#query">Query GLIDdb</a></li>
                  <li><a href="#search">Search options</a></li>
                  <li><a href="#tax">Specie/Taxonomy filter</a></li>
                  <li><a href="#pubmed">Query to PubMed</a></li>
                  <li><a href="#json">Download/Upload your results</a></li>
          </ul>
            </li>

            <li><a href="contact.html">Contact</a></li>
          </ul>
        </div><!--/.nav-collapse -->
      </div>
    </nav>

    <div class="container">
      <br><br><br>
      <section><h2 id="description">Gene to Literature Integrative Database (GLID)</h2></section><br>

      <p>Imagine you are <b>interested in a gene</b> or protein. <b>People call genes and proteins in different ways</b>, which is a nightmare for anyone
          who is interested in studying its function, evolution or clinical implications. In addition, phylogenetically related genes (potentially interesting
          if you want to study the function or evolution of the gene across organisms) do not always share nomenclature. 
          <br><br>
          At the end you may query PubMed in a partial way, missing information you would like to have.  <b>GLID aims to solve this nightmare</b>.
          <br><br>
          We have mined nomenclature data of many sources (HGNC, UniProt, NCBI, OrthoDB and others, using <a href="./Code.zip">this code</a>) and built an integrative database with this information. 
          We have provided a user-friendly interface to navigate across dfferent gene/protein names, select them and query pubmed. The database offers the possibility to 
          add keywords to the search and displaying the literature matching your patterns.
          <br><br>
          Futhermore, we have reasoned that anyone interested in a gene/protein may also be interested in those <b>genes with a similar function</b> (for example
          in the same species). We have provided two sections (similar proteins and similar function) that provide an approach to this information. 
          <br><br>
         
          <ul style="list-style-type:circle">
            <li>On the one hand, two proteins that share the exact same domains are likely to have the same function, and the section "similar proteins" provides this information from the Pfam database. </li>
            <br>
            <li> On the other hand, two genes that share a given number of Gene Ontology (GO) terms are also likely to have similar function . We have provided these relationships
                for each of the functional terms (Molecular Function, Biological Process and Cellular Component) in the "similar function" section. Note that this approach
                is limited to the fact that well-studied genes have more GO-terms and that some GO-terms are very close (in the GO relationship) but they would be considered
                different in our analysis.</li>
          </ul>
          <br><br>
          <i>NOTE: This is a pilot database, which means that only a reduced number of species (see below) has been included. This has an important impact in the 
          database content of the phylogenetically-related genes, obtained from OrthoDB. Many genes won't have any such partners because we have not included their
          corresponding phylogenetically related species. Keep this in mind when doing your search. </i>
          <br><br>
      
      </p>

      <section><h2 id="query">How to query GLIDdb</h2></section><br>
      
      <h4><i class="fa fa-search" aria-hidden="true" style="font-size:18px"></i>&nbsp;&nbsp;Enter your query</h4>
      <p>Different <b>types of query</b> are supported by GLID. In the query form, the following <b>terms</b> can be introduced:
        <ul style="list-style-type:circle">
          <li>a gene name (short or long, ether being the reccomended or an alternative name)</li>
          <li>a protein name (short or long or an alternative name)</li>
          <li>an EntrezGene ID or a UniProt ID</li>
        </ul>
        <i class="fa fa-exclamation-circle" aria-hidden="true" style="font-size:14px"></i>&nbsp;The search is <u>case-insensitive</u>, meaning that upper and lower case will be ignored and treated as equal.<br><br>
        Here there are some <b>examples</b> of usage: CDK2, cdk2, Cyclin-dependent kinase 2, 1017, P24041.</p><br>
    
      <section><h2 id="search">Search options</h2></section><br>

      <h4><i class="fa fa-check-square-o" aria-hidden="true" style="font-size:16px"></i>&nbsp;&nbsp;Look for the <b>recommended gene/protein name</b></h4>
      <p>Enter your query gene/protein name or ID, and obtain the recommended gene/protein name, if there is one. The list of recommended names are retrieved from the UniProt consortium and HGNC (HUGO Gene Nomenclature Commitee).</p>
      <p><i class="fa fa-asterisk" aria-hidden="true" style="font-size:8px"></i> From that list, you can select as many as you wish and then proceed to download a JSON file with them and/or query PubMed to maximise your literature search results.</p><br>

      <h4><i class="fa fa-check-square-o" aria-hidden="true" style="font-size:16px"></i>&nbsp;&nbsp;Find all the <b>synonyms</b> of your gene/protein</h4>
      <p>Enter your query gene/protein name or ID, and obtain a list of all the gene/protein synonyms. The list of all synonyms is retrieved from the UniProt consortium. </p>
      <p><i class="fa fa-asterisk" aria-hidden="true" style="font-size:8px"></i> From that list, you can select as many as you wish and then proceed to download a JSON file with them and/or query PubMed to maximise your literature search results.</p><br>

      <h4><i class="fa fa-check-square-o" aria-hidden="true" style="font-size:16px"></i>&nbsp;&nbsp;Find <b>phylogenetically related genes</b> from OrthoDB</h4>
      <p>Enter your query gene/protein name or ID and a specie to retrieve a list of all phylogenetically related genes annotated inside an OrthoDB group. For a better clarity, we provide a list of genes that are phylogenetically related to the query, as well as which are the species that include each of these genes.
          <a href="http://www.orthodb.org/?page=help" target="blank">Go to the OrthoDB  help page</a> for more information about how OrthoDB clusters phylogenetically related genes.
          <br><br>
          As a consequence of selecting a reduced number of species for this pilot version, only approximately a 10% of the genes are showing philogenetically related genes.</p>
      <p><i class="fa fa-asterisk" aria-hidden="true" style="font-size:8px"></i> From that list, you can select as many as you wish and then proceed to download a JSON file with them and/or query PubMed to maximise your literature search results.</p><br>

      <h4><i class="fa fa-check-square-o" aria-hidden="true" style="font-size:16px"></i>&nbsp;&nbsp;Find <b>similar proteins</b> according to PFAM domains</h4>
      <p>Enter your query gene/protein name or ID, and retrieve a list of all the described proteins which have the exact same PFAM domains. </p>
      <p><i class="fa fa-asterisk" aria-hidden="true" style="font-size:8px"></i> From that list, you can select as many as you wish and then proceed to download a JSON file with them and/or query PubMed to maximise your literature search results.</p><br>

      <h4><i class="fa fa-check-square-o" aria-hidden="true" style="font-size:16px"></i>&nbsp;&nbsp;Find proteins with <b>similar function</b> according Gene Ontology (GO)</h4>
      <p>Enter your query gene/protein name or ID, and retrieve a list of all the described proteins which, for each of the different types of Gene Ontology terms (Cellular Component, Biological Process, Molecular Function), have at least a 80% of intersection with the query.</p>
      <p><i class="fa fa-asterisk" aria-hidden="true" style="font-size:8px"></i> From that list, you can select as many as you wish and then proceed to download a JSON file with them and/or query PubMed to maximise your literature search results.</p><br>


      <section><h2 id="tax">Restrict to one specie name or taxonomy group</h2></section><br>
      <p>GLID allows to select either a <b>single specie</b> or a <b>taxonomic group</b> from two predefined sets: the specie names list and the taxonomic groups list. Regarding the specie names list, 14 species were choosen for this pilot web application. Regarding the taxonomic groups list, the taxonomic hierarchical classification division defined by the NCBI was used.<br><br>
      <i class="fa fa-exclamation-circle" aria-hidden="true" style="font-size:14px"></i>&nbsp;In the corresponding forms, start <u>typing</u> to see if your specie or taxonomic group of interest is supported.</p><br>
      <p>Here we provide the two <b>complete lists</b>:<br></p>
      <p><b>Species</b> 
        <ul>
            <li><i>Mus musculus</i></li>
            <li><i>Saccharomyces cerevisiae</i></li>
            <li><i>Schizosaccharomyces pombe</i></li>
            <li><i>Pan troglodytes</i></li>
            <li><i>Bacillus subtilis</i></li>
            <li><i>Gallus gallus</i></li>
            <li><i>Xenopus laevis</i></li>
            <li><i>Nicotiana tabacum</i></li>
            <li><i>Escherichia coli</i></li>
            <li><i>Caenorhabditis elegans</i></li>
            <li><i>Drosophila melanogaster</i></li>
            <li><i>Danio rerio</i></li>
            <li><i>Bos taurus</i></li>
            <li><i>Homo sapiens</i></li>
          </ul>
          
      <p><b>Taxonomic groups</b>
        <ul>
            <li><i>Bacteria</i></li>
            <li><i>Invertebrates</i></li>
            <li><i>Mammals</i></li>
            <li><i>Plants and Fungi</i></li>
            <li><i>Primates</i></li>
            <li><i>Rodents</i></li>
            <li><i>Rodents</i></li>
        </ul>
    <br>
    <section><h2 id="pubmed">Query to PubMed</h2></section><br>
    <p>Once your results have been reported considering your previously selected search options, you can choose <b>as many as</b> these <b>gene/protein names</b> (short and long terms) to be submitted on <b>PubMed</b>. Additionally, before submitting them, you can add some extra <b>keywords</b> if necessary. </p><br>

    <section><h2 id="json">Download/Upload your results</h2></section><br>
    <p><b>JSON file download</b><br>
      Once your results have been reported considering your previously selected search options, you can select as many as you wish and then proceed to download a JSON file.<br><br>
      <b>JSON file upload</b><br>
      At the home page, if you have <u>previously</u> performed a search and download its associated JSON array file, you can upload this file to directly submit its content to PubMed.
    </p>
    </div><!-- /.container -->

<br><br>

    <div class="container">
  <h3> <i class="fa fa-info-circle" aria-hidden="true" style="font-size:24px"></i> Help Table</h3>
  <p>Information about different concepts, webpages and databases used to build GLID.</p>            
  <table class=table table-bordered>
    <thead>
      <tr>
        <th>Name</th>
        <th>Description</th>
        <th>Link</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td><b>Names & Taxonomy</b> from <b>Uniprot</b></td>
        <td>Information on how Uniprot classifies Recommended and Alternative names</td>
        <td><a href="https://www.uniprot.org/help/names_and_taxonomy_section" target="blank">Go to UniProt</a></td>
      </tr>
      <tr>
        <td><b>OrthoDB</b> help page</td>
        <td>Information about OrthoDB parameters</td>
        <td><a href="http://www.orthodb.org/?page=help" target="blank">Go to OrthoDB help page</a></td>
      </tr>
      <tr>
          <td><b>Pfam</b> help page</td>
          <td>Information about Pfam database</td>
          <td><a href="https://pfam.xfam.org/help" target="blank">Go to Pfam help page</a></td>
        </tr>
        <tr>
            <td><b>GO</b> documentation page</td>
            <td>Information about Gene Ontology (GO) Consortium</td>
            <td><a href="http://www.geneontology.org/page/introduction-go-resource" target="blank">Go to GO documentation page</a></td>
          </tr>
      <tr>
        <td><b>PubMed</b> search help</td>
        <td>Information on how to search PubMed</td>
        <td><a href="https://www.ncbi.nlm.nih.gov/books/NBK3827/#_pubmedhelp_PubMed_Quick_Start_" target="blank">Go to PubMed Search tutorial</a></td>
      </tr>
    </tbody>
  </table>
</div>
<br>
<br>

<!-- Bootstrap core JavaScript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
    <script src="../../dist/js/bootstrap.min.js"></script>
    <!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->
    <script src="../../assets/js/ie10-viewport-bug-workaround.js"></script>

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>

  </body>
</html>
